src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,245 @@
+  }
+
+  return out;
+}
+
+// ____________________________________________________________________________
+std::ostream& operator<<(std::ostream& os, const LocatedTriple& lt) {
+  os << "LT(" << lt.blockIndex_ << " " << lt.triple_ << " "
+     << lt.shouldTripleExist_ << ")";
+  return os;
+}
+
+// ____________________________________________________________________________
+NumAddedAndDeleted LocatedTriplesPerBlock::numTriples(size_t blockIndex) const {
+  // If no located triples for `blockIndex_` exist, there is no entry in `map_`.
+  if (!map_.contains(blockIndex)) {
+    return {0, 0};
+  }
+
+  auto blockUpdateTriples = map_.at(blockIndex);
+  size_t countInserts = std::ranges::count_if(
+      blockUpdateTriples, &LocatedTriple::shouldTripleExist_);
+  return {countInserts, blockUpdateTriples.size() - countInserts};
+}
+
+// ____________________________________________________________________________
+IdTable LocatedTriplesPerBlock::mergeTriples(size_t blockIndex,
+                                             const IdTable& block,
+                                             size_t numIndexColumns) const {
+  // This method should only be called if there are located triples in the
+  // specified block.
+  AD_CONTRACT_CHECK(map_.contains(blockIndex));
+
+  // TODO<qup42>: We're assuming that the index columns are always {0, 1, 2},
+  // {1, 2} or {2}. Is this true?
+  AD_CONTRACT_CHECK(numIndexColumns <= block.numColumns());
+  AD_CONTRACT_CHECK(block.numColumns() >= 1);
+
+  auto numInsertsAndDeletes = numTriples(blockIndex);
+  IdTable result{block.numColumns(), block.getAllocator()};
+  result.resize(block.numRows() + numInsertsAndDeletes.numAdded);
+
+  const auto& locatedTriples = map_.at(blockIndex);
+
+  auto cmpLt = [&numIndexColumns](auto lt, auto row) {
+    if (numIndexColumns == 3) {
+      return std::tie(row[0], row[1], row[2]) > std::tie(lt->triple_.ids_[0],
+                                                         lt->triple_.ids_[1],
+                                                         lt->triple_.ids_[2]);
+    } else if (numIndexColumns == 2) {
+      return std::tie(row[0], row[1]) >
+             std::tie(lt->triple_.ids_[1], lt->triple_.ids_[2]);
+    } else {
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return std::tie(row[0]) > std::tie(lt->triple_.ids_[2]);
+    }
+  };
+  auto cmpEq = [&numIndexColumns](auto lt, auto row) {
+    if (numIndexColumns == 3) {
+      return std::tie(row[0], row[1], row[2]) == std::tie(lt->triple_.ids_[0],
+                                                          lt->triple_.ids_[1],
+                                                          lt->triple_.ids_[2]);
+    } else if (numIndexColumns == 2) {
+      return std::tie(row[0], row[1]) ==
+             std::tie(lt->triple_.ids_[1], lt->triple_.ids_[2]);
+    } else {
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return std::tie(row[0]) == std::tie(lt->triple_.ids_[2]);
+    }
+  };
```

Comment:
This does not work as the lambda would have different return types depending on the branch.

Reply:
Ah I understand...
I dislike this runtimeSwitch in the innermost loop since the beginning.
If you write a templated impl-function for this merging (templated on the number of index columns), and then just have an outer switch function,
then all this (trhe performance implications + the code duplication) go away.

Reply:
That would work. But I was also too focused on the individual threads to see that we don't need this at all. We said that we always require all index columns for merging, so this is obsolete anyways.